### ISSUE TITLE
derivative for channel_shuffle is not implemented

### ISSUE THREAD

**Issue by Sun-WeiZhen** [pytorch/pytorch#67240]
### Please note that this issue tracker is not a help form and this issue will be closed.

I run the program on Windows 10 using pytorch 1.10-cpu version.

I used NN. Channelshuffle (2)

Some sequences are as follows:
`        self.layer4 = nn.Sequential(
            nn.MaxPool2d(kernel_size=2, stride=2),
            nn.ChannelShuffle(2)
        )`
My error message is:

RuntimeError: derivative for channel_ shuffle is not implemented

How to solve it

- [Discussion Forum](https://discuss.pytorch.org/)

**Comment by dzhulgakov** [comment]
So it indeed seems to be a missing functionality. When channel_shuffle was added in #36815 it missed adding the derivative. Given it's a composite op by itself on cpu today, I believe it's sufficient to make the non-quantized kernel registered as `CompositeImplicitAutograd` instead of `CPU` here: https://github.com/pytorch/pytorch/blob/4e21d77dbb86581ba79d8b8179fb61a4e153e2e3/aten/src/ATen/native/native_functions.yaml#L3331

It will also add a CUDA implementation for this op which is missing currently :(

@kimishpatel can you follow up here as the original PR author?

This is not a new regression though, so not targeting 1.10.1

**Comment by kimishpatel** [comment]
Thanks for the tag @dzhulgakov. I will follow up.

**Comment by albanD** [comment]
@dzhulgakov why is this 1.10.1? This is not a regression.

**Comment by ezyang** [comment]
bumping

**Comment by kimishpatel** [comment]
@ezyang not clear to me if using `CompositeImplicitAutograd` is sufficient as suggested by @dzhulgakov. There was `native_channel_shuffle` that was introduced earlier here  in PR 50247 which `CompositeImplicitAutograd` dispatch key. But that doesnt really enable decomposition of channel shuffle into differentiable ops as I get unimplemented gradient function error. If you have a dispatch table with CPU/GPU etc. keys, then I suppose `CompositeImplicitAutograd` wouldnt quite work, would it? In that case how do we enable that? Appreciate any help.

**Comment by ezyang** [comment]
>  But that doesnt really enable decomposition of channel shuffle into differentiable ops as I get unimplemented gradient function error.

What error are you getting? The function in the quoted PR looks fully differentiable to me:

```
  Tensor output_tensor =
      input_reshaped.permute({0 /* b */, 2 /* oc */, 1 /* groups */, 3})
      .contiguous()
      .reshape(self.sizes());
```

**Comment by kimishpatel** [comment]
> What error are you getting

It is differentiable. Problem is native_functions.yaml for that op has backend specific implementation. Once you have that CompositeImplicitAutograd does not work and it says the backward for this op does not exist. Maybe we can bring this up in compsability meeting? I had a chat with @bdhirsh about it.

**Comment by bdhirsh** [comment]
Yup that's right. It looks like `channel_shuffle` has a composite implementation, but mobile wants `XNNPACK` to handle it directly.

From what I understand, the current state of `XNNPACK` is that we bundle it directly into our CPU kernels and guard it with mobile if-defs. The downside is that we can't differentiate in the dispatch table between XNNPACK and anyone else using `cpu`, which means that even if `XNNPACK` doesn't use autograd, we still need to provide an autograd formula for cpu users.

Separately, there's a `native_channel_shuffle` op that's its own NativeFunction, and I'm not 100% convinced that it needs to be its own op (and can't be bundled into `channel_shuffle`)

**Comment by ezyang** [comment]
OK, well then that just sounds like someone needs to write a backwards for it. It is probably not too difficult to do, you just need to do the inverse shuffle.

**Comment by GuillaumeTong** [comment]
I just ran into this issue on pytorch 2.1.2. Is this being worked on?
As a workaround for now, I'm using the following, with it having the benefit of also working for any other dimension of input:
``` python
x.unflatten(1, (self.groups, -1)).transpose(1, 2).contiguous().flatten(1, 2)
```
This is based on @ezyang 's earlier comment:
> ```
>   Tensor output_tensor =
>       input_reshaped.permute({0 /* b */, 2 /* oc */, 1 /* groups */, 3})
>       .contiguous()
>       .reshape(self.sizes());
> ```
